### PR TITLE
Make user.ext.prebid.buyeruids bidder name case insensitive

### PR DESCRIPTION
**Pull request by Sonali-More-Xandr**
We need to handle case insensitivity for user.ext.prebid.buyeruids biddernames. This PR updates the biddername look up to use the normalized names for buyeruids.BIDDER.

partially resolves https://github.com/prebid/prebid-server/issues/2400

### CODE REVIEW

**Review by onkarvhanumante — DISMISSED** [review]

**Comment by Sonali-More-Xandr** [comment]
@bsardo We need this PR to go out with PBS 2.0. Can you please help me understand me why this PR is marked blocked and if there is something pending from my side

**Comment by bsardo** [comment]
Hi @Sonali-More-Xandr. We originally blocked this because we thought it conflicted with #3107 and we wanted to see that merged first, but on second look it appears that was a mistake so I'll remove the label.

File: endpoints/openrtb2/amp_auction_test.go [hunk 1]
```diff
@@ -49,6 +49,7 @@ func TestGoodAmpRequests(t *testing.T) {
 				"imp-with-stored-resp.json",
 				"gdpr-no-consentstring.json",
 				"gdpr.json",
+				"aliased-buyeruids-case-insensitive.json",
```
Comment (gargcreation1992) [inline endpoints/openrtb2/amp_auction_test.go:52]: I see that you have two test case files that you added for this PR. Would it mean that it also needs to be added in this function?

File: endpoints/openrtb2/sample-requests/amp/valid-supplementary/aliased-buyeruids-camel-case.json [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+            {
+              "w": 300,
+              "h": 600
+            }
+          ]
+        },
+        "ext": {
+          "prebid": {
+            "bidder": {
+              "YahooAds": {
```
Comment (gargcreation1992) [inline endpoints/openrtb2/sample-requests/amp/valid-supplementary/aliased-buyeruids-camel-case.json:32]: Also add a test case for the request that has an alias in the `request.ext.prebid.aliases` field.

File: endpoints/openrtb2/amp_auction_test.go (hunk 1 above)
Comment (Sonali-More-Xandr) [reply endpoints/openrtb2/amp_auction_test.go:52]: Good catch! Added it in [07427d6](https://github.com/prebid/prebid-server/pull/3152/commits/07427d696b9bd7d5e38c1fc3915965d69cb34fc5)

File: endpoints/openrtb2/sample-requests/amp/valid-supplementary/aliased-buyeruids-camel-case.json (hunk 2 above)
Comment (Sonali-More-Xandr) [reply endpoints/openrtb2/sample-requests/amp/valid-supplementary/aliased-buyeruids-camel-case.json]: Added in [b43a4cf](https://github.com/prebid/prebid-server/pull/3152/commits/b43a4cfc1feeba7a0586fafd49890437b083f375)

**Review by onkarvhanumante — APPROVED** [review]

File: endpoints/openrtb2/auction.go [hunk 3]
```diff
@@ -1816,7 +1816,9 @@ func (deps *endpointDeps) validateUser(req *openrtb_ext.RequestWrapper, aliases
 			return append(errL, errors.New(`request.user.ext.prebid requires a "buyeruids" property with at least one ID defined. If none exist, then request.user.ext.prebid should not be defined.`))
 		}
 		for bidderName := range prebid.BuyerUIDs {
-			if _, ok := deps.bidderMap[bidderName]; !ok {
+			normalizedCoreBidder, _ := openrtb_ext.NormalizeBidderName(bidderName)
+			coreBidder := normalizedCoreBidder.String()
+			if _, ok := deps.bidderMap[coreBidder]; !ok {
```
Comment (VeronikaSolovei9) [inline endpoints/openrtb2/auction.go:1821]: Nitpick: in this PR here https://github.com/prebid/prebid-server/pull/3187/files#diff-785f5ab5ebde657757f3b3c34bf30a939782678d61bf1ea43f844c4f22a012d9R155:~:text=normalizeBidderName%20%20%20%20%20%20%20normalizeBidderName @SyntaxNode introduced a new function in endpoint deps. 
This validation function already has deps. Could you please pull latest changes from main branch and call normalize bidder name through the `normalizedCoreBidder, _ := deps.normalizeBidderName(bidderName)`. With this change code will be more consistent. 

Other than this looks good, local testing looks good too.

Comment (Sonali-More-Xandr) [reply endpoints/openrtb2/auction.go]: Thank you, Veronika! Updated it in [49a8d0e](https://github.com/prebid/prebid-server/pull/3152/commits/49a8d0e5cbca2fb95c4f8af2b65f5868d0cd54f4)

**Comment by VeronikaSolovei9** [comment]
Sonali, thank you for your updates! 
I want to apologize. At the moment I tested your PR at the first time, I misunderstood how this should work. 
(I did the same mistake for my StoredBidResponses PR and had to fix it... )

Here is a desired behavior.
Let's say we have a request with one imp. In the imp.ext we have this config: 
```json
"ext": {
                        "prebid": {
                            "bidder": {
                                "APPnexus": {
                                    "placementId": 1
                                },
                                "appNEXUS": {
                                    "placementId": 2
                                },
                                "amx": {}
                            }
                        }
                    }
```
We also have this config in user.ext.prebid.buyerids:
```json
"user": {
                "id": "userId",
                "ext": {
                    "prebid": {
                        "buyeruids": {
                            "APPnexUS": "12345"
                        }
                    }
                }
            }
```
in this case  "buyeruids": {"APPnexUS": "12345"} should be applied to both "APPnexus" and "appNEXUS", because we need to support case-insesitive comparisons, not case sensitive as it is now. 
My mistake was to check that   user.ext.prebid."buyeruids": { "APPnexus": "12345" } will be only applied to "APPnexus". This is not right, it should be applied to any "appnexus" in imp.ext. 

In order to fix this, add next change on top of existing ones in this PR: util.go -> func prepareUser should be changed to this:
```go
func prepareUser(req *openrtb2.BidRequest, givenBidder, syncerKey string, explicitBuyerUIDs map[string]string, usersyncs IdFetcher) bool {
	cookieId, hadCookie, _ := usersyncs.GetUID(syncerKey)

	buyerIdSet := false
	for bidderName, id := range explicitBuyerUIDs {
		if strings.EqualFold(bidderName, givenBidder) {
			req.User = copyWithBuyerUID(req.User, id)
			buyerIdSet = true
		}
	}
	if !buyerIdSet && hadCookie {
		req.User = copyWithBuyerUID(req.User, cookieId)
	}

	return hadCookie
}
```

Also I would like you to add a test that explicitly tests what data goes to what bidder.  Put this test in `exchange/exchangetest/` directory. This test has 3 bidders in imp.ext. 2 of them are "apnexus" in different casing and one amx. It also has a req.user.ext.prebid.buyerids{"APPnexUS": "12345"}.
In `outgoingRequests` there are 3 requests we expect to be executed: APPnexus, appNEXUS and amx. Only APPnexus and appNEXUS will receive user.buyeruid: "12345" -> [test_buyerids.json](https://github.com/prebid/prebid-server/files/12886333/buyerids.json)

**Comment by onkarvhanumante** [comment]
master code:
>```go
> func prepareUser(req *openrtb2.BidRequest, givenBidder, syncerKey string, explicitBuyerUIDs map[string]string, usersyncs IdFetcher) bool {
>	cookieId, hadCookie, _ := usersyncs.GetUID(syncerKey)
>
>	if id, ok := explicitBuyerUIDs[givenBidder]; ok {
>		req.User = copyWithBuyerUID(req.User, id)
>	} else if hadCookie {
>		req.User = copyWithBuyerUID(req.User, cookieId)
>	}
>
>	return hadCookie
>}


suggested code https://github.com/prebid/prebid-server/pull/3152#issuecomment-1760230797

> ```go
> func prepareUser(req *openrtb2.BidRequest, givenBidder, syncerKey string, explicitBuyerUIDs map[string]string, usersyncs IdFetcher) bool {
> 	cookieId, hadCookie, _ := usersyncs.GetUID(syncerKey)
> 
> 	buyerIdSet := false
> 	for bidderName, id := range explicitBuyerUIDs {
> 		if strings.EqualFold(bidderName, givenBidder) {
> 			req.User = copyWithBuyerUID(req.User, id)
> 			buyerIdSet = true
>                          break 
> 		}
> 	}
> 	if !buyerIdSet && hadCookie {
> 		req.User = copyWithBuyerUID(req.User, cookieId)
> 	}
> 
> 	return hadCookie
> }
> ```
> 

Thanks @VeronikaSolovei9  for above change. But something to note is this will change complexity of prepareUser from 0(1) to 0(N). Also `prepareUser()` is called within the loop  then complexity goes 0(N^2)

https://github.com/prebid/prebid-server/blob/1944dbb8592872a712532abaf7a331b2a862b2ac/exchange/utils.go#L331-L364

How about preparing lowercase key map from `explicitBuyerUIDs` and in `prepareUser()`. we could lookup using `strings.ToLower()`. Preparing lowercase key map will incur less complexity  - `0(N) + 0(N) + 0(1)`  than looping inside `prepareUser()` 



```go
lowerCaseExplicitBuyerUIDs = ... code to have lower case bidder name keys from explicitBuyerUIDs map

 for bidder, imps := range impsByBidder { 
 	coreBidder := resolveBidder(bidder, aliases) 
  
 	reqCopy := *req.BidRequest 
 	reqCopy.Imp = imps 
  
 	sChainWriter.Write(&reqCopy, bidder) 
        ...
        if hadSync := prepareUser(&reqCopy, bidder, syncerKey, explicitBuyerUIDs, auctionRequest.UserSyncs); !hadSync && req.BidRequest.App == nil { 
...

func prepareUser(req *openrtb2.BidRequest, givenBidder, syncerKey string, lowerCaseExplicitBuyerUIDs map[string]string, usersyncs IdFetcher) bool {
	cookieId, hadCookie, _ := usersyncs.GetUID(syncerKey)

	if id, ok := lowerCaseExplicitBuyerUIDs [strings.ToLower(givenBidder)]; ok {
		req.User = copyWithBuyerUID(req.User, id)
	} else if hadCookie {
		req.User = copyWithBuyerUID(req.User, cookieId)
	}

	return hadCookie
}
```

**Comment by VeronikaSolovei9** [comment]
Sure, if this gives the same result! As long as attached test passes and data is returned as expected.

**Review by VeronikaSolovei9 — APPROVED** [review]

**Review by onkarvhanumante — APPROVED** [review]

**Pull request merged**